src/puppetlabs/puppetdb/cli/benchmark.clj

```diff
@@ -123,27 +123,35 @@
                                    (repeatedly children #(random-structured-fact (rand-nth (range depth))
                                                                                  (rand-nth (range children)))))})))
 
+(defn update-catalog [catalog stamp]
+  (assoc catalog "producer_timestamp" stamp))
+
 (defn maybe-tweak-catalog
   "Slightly tweak the given catalog, returning a new catalog, `rand-percentage`
-  percent of the time."
-  [catalog rand-percentage]
-  (if (< (rand 100) rand-percentage)
-    (rand-catalog-mutation catalog)
-    catalog))
-
-(defn update-catalog [catalog]
-  (assoc catalog "producer_timestamp" (time/now)))
+   percent of the time."
+  [catalog rand-percentage stamp]
+  (let [catalog' (update-catalog catalog stamp)]
+    (if (< (rand 100) rand-percentage)
+      (rand-catalog-mutation catalog')
+      catalog')))
+
+(defn jitter
+  "jitter a timestamp (rand n) milliseconds in the forward direction"
+  [stamp n]
+  (time/plus stamp (time/seconds (rand n))))
 
 (defn update-report-run-fields
   "configuration_version, start_time and end_time should always change
    on subsequent report submittions, this changes those fields to avoid
    computing the same hash again (causing constraint errors in the DB)"
-  [report]
+  [report stamp]
   (assoc report
-         "configuration_version" (ks/uuid)
-         "start_time" (time/now)
-         "end_time" (time/now)
-         "producer_timestamp" (time/now)))
+    "configuration_version" (ks/uuid)
+    "resource_events" (map #(assoc % "timestamp" (jitter stamp 300))
```

Comment:
I know what you mean about the ugliness of the `update` when it's a list of things, it would be nice if the there was an `update-coll` function that was a wrapper around `update ... (partial map`, but I still think at the top level here you're updating and at the `map` level you're overwriting.